File: Depth_First_Search/1443.Minimum_Time_to_Collect_All_Apples_in_a_Tree/C++/1443.Minimum_Time_to_Collect_All_Apples_in_a_Tree.cpp
class Solution {
public:
    int minTime(int n, vector<vector<int>> & edges, vector<bool> & hasApple) 
    {
        //=============================================//
        // Input type:                                 //
        //   - int (n-th nodes)                        //
        //   - vector<vector<int>> ref. (edges vector) //
        //   - vector<bool> ref. (hasApple vector)     //
        // Return type:                                //
        //   - int (result minimun times)              //
        //=============================================//

        /*Initialize*/
        ///// Length of edges vector /////
        int len_edges = edges.size();

        ///// Root-child dictionary, hashmap /////
        unordered_map<int, vector<int>> root_child_map;

        ///// Visit vector /////
        vector<bool> visit_vect(n, false);


        /*Recursion-based DFS loop traversal with recorded dictionary, hashmap*/
        visit_vect[0] = true; //Update, Overwrite

        for (int edges_idx = 0; (edges_idx < len_edges); (edges_idx++)) //Whole
        {
            (root_child_map[(edges[edges_idx])[0]]).emplace_back((edges[edges_idx])[1]); //Keep updating, recording
            (root_child_map[(edges[edges_idx])[1]]).emplace_back((edges[edges_idx])[0]); //Keep updating, recording

        } //Whole

        return dfsTraversal(root_child_map, 0, visit_vect, hasApple);
    }


    int dfsTraversal(unordered_map<int, vector<int>> & root_child_map, int record_node, 
                     vector<bool> & visit_vect, const vector<bool> & hasApple)
    {
        //===========================================================================//
        // Input type:                                                               //
        //   - unordered_map<int, vector<int>> ref. (root-child dictionary, hashmap) //
        //   - int (record node)                                                     //
        //   - vector<bool> ref. (visit vector)                                      //
        //   - const vector<bool> ref. (hasApple vector)                             //
        // Return type:                                                              //
        //   - int (result minimun times)                                            //
        //===========================================================================//

        /*Initialize*/
        ///// Result summary counter /////
        int res_cnter = 0;

        /*Whole process, flow*/
        for (int & next_node: root_child_map[record_node]) //Whole
        {
            ///// Check if the next node existed or not /////
            if (! visit_vect[next_node])
            {
                visit_vect[next_node] = true; //Keep updating, overwriting

                int record_cnter = dfsTraversal(root_child_map, next_node, visit_vect, hasApple); //Record summary counter

                ///// Check if the current status matched conditions or not /////
                if (record_cnter || hasApple[next_node]) { res_cnter += (record_cnter + 2); } //Keep updating, accumulating
            }

        } //Whole

        return res_cnter;
    }
};

File: Depth_First_Search/1443.Minimum_Time_to_Collect_All_Apples_in_a_Tree/Python/1443.Minimum_Time_to_Collect_All_Apples_in_a_Tree.py
class Solution(object):
    def minTime(self, n, edges, hasApple):
        """
        :type n: int
        :type edges: List[List[int]]
        :type hasApple: List[bool]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of edges array #####
        len_edges = len(edges)

        ##### Root-child dictionary/hashmap #####
        root_child_dict = {}

        ##### Visit array #####
        visit_arry = [False for _ in range(n)]


        #####################################################################
        #Recursion-based DFS loop traversal with recorded dictionary/hashmap
        visit_arry[0] = True #Update/Overwrite

        for edges_idx in range(len_edges):
            curr_node, next_node = (edges[edges_idx])[0], (edges[edges_idx])[1] #Current, Next node

            ##### Check if the current node existed or not #####
            if (curr_node not in root_child_dict):
                root_child_dict[curr_node] = [next_node] #Keep updating/recording
            else:
                (root_child_dict[curr_node]).append(next_node) #Keep updating/recording

            ##### Check if the next node existed or not #####
            if (next_node not in root_child_dict):
                root_child_dict[next_node] = [curr_node] #Keep updating/recording
            else:
                (root_child_dict[next_node]).append(curr_node) #Keep updating/recording

        return self.dfsTraversal(root_child_dict, 0, visit_arry, hasApple)

        
    def dfsTraversal(self, root_child_dict, record_node, visit_arry, hasApple):
        """
        :type root_child_dict: dict
        :type record_node: int
        :type visit_arry: List[bool]
        :type hasApple: List[bool]
        :rtype: int
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#

        ############
        #Initialize
        ##### Result summary counter #####
        res_cnter = 0

        ####################
        #Whole process/flow
        if (record_node in root_child_dict):
            for next_node in root_child_dict[record_node]:

                ##### Check if the next node is visited or not #####
                if (not visit_arry[next_node]):
                    visit_arry[next_node] = True #Keep updating/overwriting

                    record_cnter = self.dfsTraversal(root_child_dict, next_node, visit_arry, hasApple) #Record summary counter

                    ##### Check if the current status matched conditions or not #####
                    if (record_cnter or hasApple[next_node]):
                        res_cnter += (record_cnter + 2) #Keep updating/accumulating

        return res_cnter
